mbf-site/src/App.tsx

```diff
@@ -71,12 +78,126 @@ export async function getAndroidVersion(device: Adb) {
   return Number(result.stdout.trim());
 }
 
+function areDevicesEqual(devices1: Record<string, any>[], devices2: Record<string, any>[]): boolean {
+  if (devices1.length !== devices2.length) {
+    return false;
+  }
+
+  for (let i = 0; i < devices1.length; i++) {
+    const device1 = devices1[i];
+    const device2 = devices2[i];
+
+    if (!areObjectsEqual(device1, device2)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+function areObjectsEqual(obj1: Record<string, any>, obj2: Record<string, any>): boolean {
+  const keys1 = Object.keys(obj1);
+  const keys2 = Object.keys(obj2);
+
+  if (keys1.length !== keys2.length) {
+    return false;
+  }
+
+  for (const key of keys1) {
+    if (obj1[key] !== obj2[key]) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
 function ChooseDevice() {
   const [authing, setAuthing] = useState(false);
   const [chosenDevice, setChosenDevice] = useState(null as Adb | null);
   const [connectError, setConnectError] = useState(null as string | null);
   const [devicePreV51, setdevicePreV51] = useState(false);
   const [deviceInUse, setDeviceInUse] = useState(false);
+  const [checkedForBridge, setCheckedForBridge] = useState(false);
+  const [bridgeClient, setBridgeClient] = useState<AdbServerClient | null>(null);
+  const [adbDevices, setAdbDevices] = useState<AdbServerClient.Device[]>([]);
+
+  // Check if the bridge is running
+  useEffect(() => {
+    if (!checkedForBridge) {
+      checkForBridge().then(haveBridge => {
+        if (haveBridge) {
+          setBridgeClient(new AdbServerClient(new AdbServerWebSocketConnector()));
+        }
+
+        setCheckedForBridge(true);
+      });
+    }
+  });
+
+  // Update the available devices on an interval
+  useEffect(() => {
```

Comment:
This component is definitely getting to the size where it should be broken up a bit in my opinion.